File: client/src/ipc/messages.rs
use serde::{Deserialize, Serialize, de::DeserializeOwned};
use serde_json::{Value, json};
use uuid::Uuid;

/// All IPC message types exchanged between the Orion runtime and clients.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageType {
    // ----------------------------------------------------------------------
    // Connection
    // ----------------------------------------------------------------------

    Ping,
    Pong,

    // ----------------------------------------------------------------------
    // Prompt submission
    // ----------------------------------------------------------------------

    SubmitPrompt,
    CancelRequest,

    // ----------------------------------------------------------------------
    // Voice
    // ----------------------------------------------------------------------

    VoiceStart,
    VoiceChunk,
    VoiceEnd,

    // ----------------------------------------------------------------------
    // Assistant streaming
    // ----------------------------------------------------------------------

    AssistantStart,
    AssistantChunk,
    AssistantEnd,

    // ----------------------------------------------------------------------
    // Tool execution
    // ----------------------------------------------------------------------

    ToolStarted,
    ToolFinished,

    // ----------------------------------------------------------------------
    // Runtime
    // ----------------------------------------------------------------------

    Status,
    Error,
}

/// Every IPC message exchanged between Orion and a client is wrapped inside
/// an Envelope.
///
/// The envelope contains transport-level metadata while the payload contains
/// the message-specific data.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Envelope {
    /// IPC protocol version.
    #[serde(default = "default_version")]
    pub version: u32,

    /// Unique identifier for this individual message.
    #[serde(default = "Uuid::new_v4")]
    pub id: Uuid,

    /// Identifier connecting related messages to the same request/pipeline.
    #[serde(default = "Uuid::new_v4")]
    pub correlation_id: Uuid,

    /// Type of message contained in the envelope.
    #[serde(rename = "type")]
    pub message_type: MessageType,

    /// Message-specific payload.
    #[serde(default)]
    pub payload: Value,
}

const fn default_version() -> u32 {
    1
}

impl Envelope {
    /// Create an envelope with a newly generated message ID and correlation ID.
    pub fn new(message_type: MessageType, payload: Value) -> Self {
        Self {
            version: 1,
            id: Uuid::new_v4(),
            correlation_id: Uuid::new_v4(),
            message_type,
            payload,
        }
    }

    /// Create an envelope using an existing correlation ID.
    ///
    /// This is important for runtime events because all events belonging to
    /// one pipeline must remain associated with the same request.
    pub fn with_correlation_id(
        message_type: MessageType,
        correlation_id: Uuid,
        payload: Value,
    ) -> Self {
        Self {
            version: 1,
            id: Uuid::new_v4(),
            correlation_id,
            message_type,
            payload,
        }
    }

    /// Create a ping request.
    pub fn ping() -> Self {
        Self::new(MessageType::Ping, json!({}))
    }

    /// Create a pong response.
    pub fn pong() -> Self {
        Self::new(MessageType::Pong, json!({}))
    }

    /// Create a prompt submission.
    pub fn submit_prompt(text: impl Into<String>) -> Self {
        Self::new(
            MessageType::SubmitPrompt,
            serde_json::to_value(SubmitPromptPayload {
                text: text.into(),
            })
            .expect("failed to serialize SubmitPromptPayload"),
        )
    }

    /// Create a prompt submission associated with an existing request.
    pub fn submit_prompt_with_correlation(
        correlation_id: Uuid,
        text: impl Into<String>,
    ) -> Self {
        Self::with_correlation_id(
            MessageType::SubmitPrompt,
            correlation_id,
            serde_json::to_value(SubmitPromptPayload {
                text: text.into(),
            })
            .expect("failed to serialize SubmitPromptPayload"),
        )
    }

    /// Create a cancel request.
    pub fn cancel_request(request_id: Uuid) -> Self {
        Self::new(
            MessageType::CancelRequest,
            serde_json::to_value(CancelRequestPayload { request_id })
                .expect("failed to serialize CancelRequestPayload"),
        )
    }

    /// Create a voice recording start message.
    pub fn voice_start(sample_rate: u32, channels: u32) -> Self {
        Self::new(
            MessageType::VoiceStart,
            serde_json::to_value(VoiceStartPayload {
                sample_rate,
                channels,
                encoding: "pcm16".to_string(),
            })
            .expect("failed to serialize VoiceStartPayload"),
        )
    }

    /// Create a voice chunk message.
    pub fn voice_chunk(sequence: u64, data: Vec<u8>) -> Self {
        Self::new(
            MessageType::VoiceChunk,
            serde_json::to_value(VoiceChunkPayload { sequence, data })
                .expect("failed to serialize VoiceChunkPayload"),
        )
    }

    /// Finish a voice recording by providing the recorded file path.
    pub fn voice_end(path: impl Into<String>) -> Self {
        Self::new(
            MessageType::VoiceEnd,
            serde_json::to_value(VoiceEndPayload { path: path.into() })
                .expect("failed to serialize VoiceEndPayload"),
        )
    }

    /// Signal that assistant response generation has started.
    pub fn assistant_start(correlation_id: Uuid) -> Self {
        Self::with_correlation_id(
            MessageType::AssistantStart,
            correlation_id,
            json!({}),
        )
    }

    /// Send an assistant response chunk.
    pub fn assistant_chunk(
        correlation_id: Uuid,
        text: impl Into<String>,
    ) -> Self {
        Self::with_correlation_id(
            MessageType::AssistantChunk,
            correlation_id,
            serde_json::to_value(AssistantChunkPayload {
                text: text.into(),
            })
            .expect("failed to serialize AssistantChunkPayload"),
        )
    }

    /// Signal that assistant response generation has completed.
    pub fn assistant_end(correlation_id: Uuid) -> Self {
        Self::with_correlation_id(
            MessageType::AssistantEnd,
            correlation_id,
            json!({}),
        )
    }

    /// Signal that a tool has started.
    pub fn tool_started(
        correlation_id: Uuid,
        name: impl Into<String>,
    ) -> Self {
        Self::with_correlation_id(
            MessageType::ToolStarted,
            correlation_id,
            serde_json::to_value(ToolStartedPayload {
                name: name.into(),
            })
            .expect("failed to serialize ToolStartedPayload"),
        )
    }

    /// Signal that a tool has finished.
    pub fn tool_finished(
        correlation_id: Uuid,
        name: impl Into<String>,
        success: bool,
    ) -> Self {
        Self::with_correlation_id(
            MessageType::ToolFinished,
            correlation_id,
            serde_json::to_value(ToolFinishedPayload {
                name: name.into(),
                success,
            })
            .expect("failed to serialize ToolFinishedPayload"),
        )
    }

    /// Send a runtime status update.
    pub fn status(
        correlation_id: Uuid,
        message: impl Into<String>,
    ) -> Self {
        Self::with_correlation_id(
            MessageType::Status,
            correlation_id,
            serde_json::to_value(StatusPayload {
                message: message.into(),
            })
            .expect("failed to serialize StatusPayload"),
        )
    }

    /// Send an error to the client.
    pub fn error(
        correlation_id: Uuid,
        code: impl Into<String>,
        message: impl Into<String>,
    ) -> Self {
        Self::with_correlation_id(
            MessageType::Error,
            correlation_id,
            serde_json::to_value(ErrorPayload {
                code: code.into(),
                message: message.into(),
            })
            .expect("failed to serialize ErrorPayload"),
        )
    }

    /// Deserialize the envelope payload into a strongly typed payload.
    pub fn payload<T>(&self) -> serde_json::Result<T>
    where
        T: DeserializeOwned,
    {
        serde_json::from_value(self.payload.clone())
    }

    /// Serialize this envelope to JSON.
    pub fn to_json(&self) -> serde_json::Result<String> {
        serde_json::to_string(self)
    }

    /// Deserialize an envelope from JSON.
    pub fn from_json(json: &str) -> serde_json::Result<Self> {
        serde_json::from_str(json)
    }
}

/* -------------------------------------------------------------------------- */
/*                             Payload Definitions                            */
/* -------------------------------------------------------------------------- */

/// User prompt submitted to Orion.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubmitPromptPayload {
    pub text: String,
}

/// Request cancellation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CancelRequestPayload {
    pub request_id: Uuid,
}

/// Metadata describing a voice recording session.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceStartPayload {
    pub sample_rate: u32,
    pub channels: u32,
    pub encoding: String,
}

/// A streamed audio frame.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceChunkPayload {
    pub sequence: u64,
    pub data: Vec<u8>,
}

/// Marks the end of a voice recording.
///
/// The runtime reads and transcribes the specified file.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceEndPayload {
    pub path: String,
}

/// Signals the start of assistant response generation.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AssistantStartPayload {}

/// A streamed fragment of assistant output.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssistantChunkPayload {
    pub text: String,
}

/// Signals completion of assistant response generation.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AssistantEndPayload {}

/// Indicates that a tool has started executing.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolStartedPayload {
    pub name: String,
}

/// Indicates that a tool has completed.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolFinishedPayload {
    pub name: String,
    pub success: bool,
}

/// General runtime status update.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StatusPayload {
    pub message: String,
}

/// Runtime error returned to the client.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorPayload {
    pub code: String,
    pub message: String,
}

/// Ping payload.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct PingPayload {}

/// Pong payload.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct PongPayload {}

File: runtime/src/orion/transport/messages.py
from __future__ import annotations

from enum import StrEnum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field


# ======================================================================
# Message Types
# ======================================================================


class MessageType(StrEnum):
    """
    All message types supported by the Orion IPC protocol.

    Transport messages are intentionally independent of Orion's internal
    domain event classes. Internal events are translated into Envelopes
    before being sent to clients.
    """

    # ------------------------------------------------------------------
    # Connection
    # ------------------------------------------------------------------

    #: Sent by a client to verify that the runtime is reachable.
    PING = "ping"

    #: Sent by the runtime in response to a ping request.
    PONG = "pong"

    # ------------------------------------------------------------------
    # Prompt Submission
    # ------------------------------------------------------------------

    #: Submit a text prompt for the assistant to process.
    SUBMIT_PROMPT = "submit_prompt"

    #: Cancel an in-progress assistant request.
    CANCEL_REQUEST = "cancel_request"

    # ------------------------------------------------------------------
    # Voice Streaming
    # ------------------------------------------------------------------

    #: Indicates the beginning of a voice recording session.
    VOICE_START = "voice_start"

    #: Contains one chunk of streamed audio.
    VOICE_CHUNK = "voice_chunk"

    #: Indicates that the client has finished sending audio.
    VOICE_END = "voice_end"

    # ------------------------------------------------------------------
    # Assistant Streaming
    # ------------------------------------------------------------------

    #: Indicates that the assistant has started generating a response.
    ASSISTANT_START = "assistant_start"

    #: Contains one streamed chunk of assistant text.
    ASSISTANT_CHUNK = "assistant_chunk"

    #: Indicates that the assistant response has completed.
    ASSISTANT_END = "assistant_end"

    # ------------------------------------------------------------------
    # Tool Execution
    # ------------------------------------------------------------------

    #: Indicates that the assistant has started executing a tool.
    TOOL_STARTED = "tool_started"

    #: Indicates that a tool execution has completed.
    TOOL_FINISHED = "tool_finished"

    # ------------------------------------------------------------------
    # Pipeline
    # ------------------------------------------------------------------

    #: Indicates that a processing pipeline has started.
    PIPELINE_STARTED = "pipeline_started"

    #: Indicates that a processing pipeline completed successfully.
    PIPELINE_COMPLETED = "pipeline_completed"

    #: Indicates that a processing pipeline failed.
    PIPELINE_FAILED = "pipeline_failed"

    #: Indicates that a processing pipeline was restarted.
    PIPELINE_RESTARTED = "pipeline_restarted"

    # ------------------------------------------------------------------
    # Runtime
    # ------------------------------------------------------------------

    #: General runtime status update.
    STATUS = "status"

    #: Indicates an error.
    ERROR = "error"

    #: Runtime event
    RUNTIME_EVENT = "runtime_event"


# ======================================================================
# Envelope
# ======================================================================


class Envelope(BaseModel):
    """
    Transport envelope exchanged between the Orion runtime and clients.

    Every message transmitted over IPC is wrapped inside an Envelope.

    The payload is interpreted according to the message type.
    """

    model_config = ConfigDict(
        use_enum_values=True,
    )

    #: Protocol version used for compatibility checks.
    version: int = 1

    #: Unique identifier for this individual message.
    id: UUID = Field(
        default_factory=uuid4,
    )

    #: Identifier of the request/workflow this message belongs to.
    correlation_id: UUID | None = None

    #: Type of message being transmitted.
    type: MessageType

    #: Message-specific payload.
    payload: dict[str, Any]


# ======================================================================
# Client -> Runtime
# ======================================================================


class SubmitPromptPayload(BaseModel):
    """A user prompt submitted to the assistant."""

    #: User's prompt.
    text: str


class CancelRequestPayload(BaseModel):
    """Request cancellation of an active assistant task."""

    #: Identifier of the request to cancel.
    request_id: UUID


class VoiceStartPayload(BaseModel):
    """Metadata describing a new voice recording session."""

    #: Audio sampling frequency.
    sample_rate: int

    #: Number of audio channels.
    channels: int

    #: Audio encoding format.
    encoding: str = "pcm16"


class VoiceChunkPayload(BaseModel):
    """A single streamed audio frame."""

    #: Sequential chunk number.
    sequence: int

    #: Raw encoded audio bytes.
    data: bytes


class VoiceEndPayload(BaseModel):
    """
    Marks the end of a voice recording.

    Carries the path to the file the runtime should process.
    """

    #: Filesystem path to the recorded audio.
    path: str


# ======================================================================
# Runtime -> Client
# ======================================================================


# ----------------------------------------------------------------------
# Assistant
# ----------------------------------------------------------------------


class AssistantStartPayload(BaseModel):
    """Signals the start of assistant response generation."""


class AssistantChunkPayload(BaseModel):
    """A streamed fragment of assistant output."""

    #: Partial assistant response.
    text: str


class AssistantEndPayload(BaseModel):
    """Signals completion of assistant response generation."""


# ----------------------------------------------------------------------
# Pipeline
# ----------------------------------------------------------------------


class PipelineStartedPayload(BaseModel):
    """Signals that a processing pipeline has started."""

    #: Name of the pipeline.
    pipeline: str


class PipelineCompletedPayload(BaseModel):
    """Signals that a processing pipeline completed successfully."""

    #: Name of the pipeline.
    pipeline: str


class PipelineFailedPayload(BaseModel):
    """
    Describes a failed processing pipeline.

    The traceback is optional so production clients do not have to
    expose implementation details while development clients can still
    display complete debugging information.
    """

    #: Name of the pipeline that failed.
    pipeline: str

    #: Machine-readable exception type.
    error_type: str

    #: Human-readable error message.
    message: str

    #: Full traceback when available.
    traceback: str | None = None


class PipelineRestartedPayload(BaseModel):
    """Signals that a processing pipeline has been restarted."""

    #: Name of the pipeline.
    pipeline: str


# ----------------------------------------------------------------------
# Tool Execution
# ----------------------------------------------------------------------


class ToolStartedPayload(BaseModel):
    """Indicates that the assistant has started executing a tool."""

    #: Name of the tool.
    name: str


class ToolFinishedPayload(BaseModel):
    """Indicates that a tool execution has completed."""

    #: Name of the tool.
    name: str

    #: Whether execution completed successfully.
    success: bool


# ----------------------------------------------------------------------
# Runtime Status
# ----------------------------------------------------------------------


class StatusPayload(BaseModel):
    """General runtime status update."""

    #: Human-readable status message.
    message: str


class ErrorPayload(BaseModel):
    """
    Represents an error returned by the runtime.

    This is intended for errors that are not represented by a more
    specific pipeline failure message.
    """

    #: Machine-readable error identifier.
    code: str

    #: Human-readable error description.
    message: str

    #: Exception type when available.
    error_type: str | None = None

    #: Pipeline associated with the error when applicable.
    pipeline: str | None = None

    #: Full traceback when available.
    traceback: str | None = None


# ----------------------------------------------------------------------
# Voice / Speech
# ----------------------------------------------------------------------


class VoiceRecordingStartedPayload(BaseModel):
    """Signals that voice recording has started."""


class VoiceRecordingCompletedPayload(BaseModel):
    """Signals that voice recording has completed."""

    #: Path to the recorded audio.
    path: str | None = None


class TranscriptPayload(BaseModel):
    """Contains a generated speech transcript."""

    #: Transcribed text.
    text: str


class SpeechSynthesisPayload(BaseModel):
    """Contains information about generated speech."""

    #: Text that was synthesized.
    text: str

    #: Generated audio path, if available.
    audio_path: str | None = None


# ----------------------------------------------------------------------
# Audio Playback
# ----------------------------------------------------------------------


class AudioPlaybackStartedPayload(BaseModel):
    """Signals that audio playback has started."""


class AudioPlaybackCompletedPayload(BaseModel):
    """Signals that audio playback has completed."""


# ----------------------------------------------------------------------
# Connection
# ----------------------------------------------------------------------


class PingPayload(BaseModel):
    """Ping request payload."""


class PongPayload(BaseModel):
    """Ping response payload."""